farmerbot/internal/farmerbot.go

```diff
@@ -0,0 +1,425 @@
+	}
+
+	// get old node state
+	nodeObj.timeoutClaimedResources = oldNode.timeoutClaimedResources
+	nodeObj.lastTimePowerStateChanged = oldNode.lastTimePowerStateChanged
+	nodeObj.lastTimeAwake = oldNode.lastTimeAwake
+	nodeObj.timesRandomWakeUps = oldNode.timesRandomWakeUps
+
+	if nodeObj.powerState == wakingUP && time.Since(nodeObj.lastTimePowerStateChanged) > timeoutPowerStateChange {
+		log.Warn().Uint32("nodeID", uint32(nodeObj.ID)).Msg("Wakeup was unsuccessful. Putting its state back to off.")
+		nodeObj.powerState = off
+		nodeObj.lastTimePowerStateChanged = time.Now()
+	}
+
+	if nodeObj.powerState == shuttingDown && time.Since(nodeObj.lastTimePowerStateChanged) > timeoutPowerStateChange {
+		log.Warn().Uint32("nodeID", uint32(nodeObj.ID)).Msg("Shutdown was unsuccessful. Putting its state back to on.")
+		nodeObj.powerState = on
+		nodeObj.lastTimeAwake = time.Now()
+		nodeObj.lastTimePowerStateChanged = time.Now()
+	}
+
+	err = f.state.updateNode(nodeObj)
+	if err != nil {
+		return node{}, err
+	}
+
+	return nodeObj, nil
+}
+
+func (f *FarmerBot) shouldWakeUp(sub Sub, node node, roundStart time.Time) bool {
+	if node.powerState != off {
+		return false
+	}
+
+	periodicWakeUpStart := f.config.Power.PeriodicWakeUpStart.PeriodicWakeUpTime()
+
+	if periodicWakeUpStart.Before(roundStart) && node.lastTimeAwake.Before(periodicWakeUpStart) {
+		// we wake up the node if the periodic wake up start time has started and only if the last time the node was awake
+		// was before the periodic wake up start of that day
+		log.Info().Uint32("nodeID", uint32(node.ID)).Msg("Periodic wake up")
+		return true
+	}
+
+	nodesLen := len(f.nodes)
+
+	// TODO:
+	if node.timesRandomWakeUps < defaultRandomWakeUpsAMonth &&
+		int(rand.Int31())%((8460-(defaultRandomWakeUpsAMonth*6)-
+			(defaultRandomWakeUpsAMonth*(nodesLen-1))/int(math.Min(float64(f.config.Power.PeriodicWakeUpLimit), float64(nodesLen))))/
+			defaultRandomWakeUpsAMonth) == 0 {
+		// Random periodic wake up (10 times a month on average if the node is almost always down)
+		// we execute this code every 5 minutes => 288 times a day => 8640 times a month on average (30 days)
+		// but we have 30 minutes of periodic wake up every day (6 times we do not go through this code) => so 282 times a day => 8460 times a month on average (30 days)
+		// as we do a random wake up 10 times a month we know the node will be on for 30 minutes 10 times a month so we can subtract 6 times the amount of random wake ups a month
+		// we also do not go through the code if we have woken up too many nodes at once => subtract (10 * (n-1))/min(periodic_wake up_limit, amount_of_nodes) from 8460
+		// now we can divide that by 10 and randomly generate a number in that range, if it's 0 we do the random wake up
+		log.Info().Uint32("nodeID", uint32(node.ID)).Msg("Random wake up")
+		return true
+	}
+
+	return false
+}
+
+func (f *FarmerBot) validateAccountEnoughBalance(sub *substrate.Substrate) error {
```

Comment:
This definitely needs to be called during the farmerbot start, we shouldn't let them start it if they don't have enough balance to maintain it running for a month at least

Reply:
As per our discussion, we will set 100 TFT as a minimum balance allowed for running the bot.